File: backend/src/main/java/com/taskmanager/backend/service/TaskService.java
package com.taskmanager.backend.service;

import com.taskmanager.backend.entity.Task;
import com.taskmanager.backend.repository.TaskRepository;
import org.springframework.stereotype.Service;

import java.util.List;
import java.util.Optional;

@Service
public class TaskService {

    private final TaskRepository taskRepository;

    // Constructor injection
    public TaskService(TaskRepository taskRepository) {
        this.taskRepository = taskRepository;
    }

    // Retrieve all tasks
    public List<Task> getAllTasks() {
        return taskRepository.findAll();
    }

    // Retrieve task by id
    public Optional<Task> getTaskById(Long id) {
        return taskRepository.findById(id);
    }

    // Create a new task
    public Task createTask(Task task) {
        return taskRepository.save(task);
    }

    // Update an existing task
    public Optional<Task> updateTask(Long id, Task taskDetails) {
        return taskRepository.findById(id).map(existingTask -> {
            existingTask.setTitle(taskDetails.getTitle());
            existingTask.setDescription(taskDetails.getDescription());
            existingTask.setStatus(taskDetails.getStatus());
            return taskRepository.save(existingTask);
        });
    }

    // Delete a task by id
    public boolean deleteTask(Long id) {
        if (taskRepository.existsById(id)) {
            taskRepository.deleteById(id);
            return true;
        }
        return false;
    }
}

File: frontend/src/services/taskService.js
import axios from 'axios';

const API_URL = 'http://localhost:8080/api/tasks';

const taskService = {
  // Fetch all tasks
  getAllTasks: async () => {
    const response = await axios.get(API_URL);
    return response.data;
  },

  // Add a new task
  createTask: async (taskData) => {
    // Expected taskData: { title, description, status: 'Pending' | 'Completed' }
    const response = await axios.post(API_URL, taskData);
    return response.data;
  },

  // Update an existing task
  updateTask: async (id, taskData) => {
    // Expected taskData: { title, description, status }
    const response = await axios.put(`${API_URL}/${id}`, taskData);
    return response.data;
  },

  // Delete a task
  deleteTask: async (id) => {
    const response = await axios.delete(`${API_URL}/${id}`);
    return response.data;
  }
};

export default taskService;
